### PR TITLE
 SCP-660: metadata for endpoints 

### PR DESCRIPTION
**Pull request by j-mueller**
* Add a field `aeMetadata :: Maybe JSON.Value` field to the `EndpointDescription` type
* This field can be used to pass arbitrary JSON values to the user when exposing an endpoint

I thought about adding a type variable for the metadata, but: (1) we already have enough type parameters, (2) we don't do anything with the metadata on the contract side, it exists only to carry information for the frontaned and (3) whatever value we put in there we will have to serialise to JSON anyway.

### CODE REVIEW

File: plutus-contract/src/Language/Plutus/Contract/Effects/ExposeEndpoint.hs [hunk 1]
```diff
@@ -47,10 +49,19 @@ type HasEndpoint l a s =
...
+  }
+  deriving (Eq, Show, Generic)
   deriving anyclass (ToJSON, FromJSON)
-  deriving Pretty via (Tagged "ActiveEndpoint:" EndpointDescription)
+
+instance Pretty ActiveEndpoint where
+  pretty ActiveEndpoint{aeDescription, aeMetadata} =
+    indent 2 $ vsep
+      [ "Endpoint:" <+> pretty aeDescription
+      , "Data:" <+> viaShow aeMetadata
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/Effects/ExposeEndpoint.hs:63]: Metadata?

**Comment by j-mueller** [comment]
> I do wonder if we should enforce some structure on the JSON. Really want nice row-types here: would be nice to say "this has to have name/description but otherwise it's up to you".

I think the arguments for/against the type parameter apply here as well. With the current definition we're saying that the JSON that you get out always has a `aeDescription` field and sometimes it also has an `aeMetadata` field. If we had some `row-types` constraint instead of `aeMetadata :: Maybe Value` the difference would at most be one level of nesting in the output.

Example:

```haskell
endpointWithRow :: forall l a s e b. (HasType l a v, HasType "name" String b, ...) => Rec (b .// "name") -> Contract s e a
endpointWithMeta :: forall l a s e b. (ToJSON b, ...) => b -> Contract s e a

endpointWithRow @"password" @_ @_ @_ @["minLength" :-> Int] ("minLength" .~ 10)
endpointWithMeta @"password" @_ @_ @_ @MinLength (MinLength 10)
```

would get you this json

```json
{"name":"password","minLength":10}
{"aeDescription":"password","aeMetadata":{"minLength":10}}
```

**Comment by michaelpj** [comment]
Hm, I see. So we *could* just say "if you want to return metadata, then add an endpoint for each bit of metadata you return which returns that.

I guess there's some value to having it in one endpoint, though: we can say "this should never have any effects", and maybe "you also get it when you start up the contract" or something.

**Comment by michaelpj** [comment]
Er, sorry, I was getting confused, this is per-endpoint metadata.

I agree that we'd just get slightly differently nested JSON, the main thing that I wanted was the ability to enforce *some* structure via types. Otherwise we have no way of really making sure that people *do* provide "name".

**Comment by j-mueller** [comment]
> the main thing that I wanted was the ability to enforce some structure via types. Otherwise we have no way of really making sure that people do provide "name".

Sorry, I don't understand if there's anything missing from the current version. At the moment we always have a "name" value, which is the value-level string of the symbol type of the endpoint label. So you can't call `endpoint` without the correct name ending up in the JSON that's produced.

**Comment by j-mueller** [comment]
We don't need to ask the user to put in the "name" value, because we put it in ourselves.

**Comment by michaelpj** [comment]
Er, sorry, "name" is a bad example since that is already in there. Let's pick some other kind of metadata we might want to enforce, e.g. "description". At the moment it's just either in the bag of JSON or not. But maybe that's fine.

**Comment by j-mueller** [comment]
I see! I misunderstood what you were saying then. Yeah in that case it would be useful to have a row type, I agree. But that should be easy to do on top of `endpointWithMeta` when we need it.

**Pull request merged**